### PR TITLE
store/tikv: implement coprocessor streaming in mocktikv

### PR DESCRIPTION
**Pull request by tiancaiamao**
So if we enable streaming API, startup a TiDB locally would not panic.

### CODE REVIEW

File: store/tikv/mocktikv/cop_handler_dag.go [hunk 1]
```diff
@@ -95,7 +101,24 @@ func (h *rpcHandler) handleCopDAGRequest(req *coprocessor.Request) *coprocessor.
...
+func (h *rpcHandler) handleCopStream(req *coprocessor.Request) (tikvpb.Tikv_CoprocessorStreamClient, error) {
+	if len(req.Ranges) == 0 {
+		return nil, errors.Errorf("empty request range")
+	}
+	switch req.Tp {
+	case kv.ReqTypeDAG:
+		return h.handleCopStreamDAG(req)
+	case kv.ReqTypeAnalyze:
+	}
+	return nil, errors.New("not implement yet")
```
Comment (zz-jason) [inline store/tikv/mocktikv/cop_handler_dag.go]: s/not implement yet/unsupported cop request type/

File: store/tikv/mocktikv/cop_handler_dag.go [hunk 2]
```diff
@@ -95,7 +101,24 @@ func (h *rpcHandler) handleCopDAGRequest(req *coprocessor.Request) *coprocessor.
...
+}
+
+func (h *rpcHandler) handleCopStreamDAG(req *coprocessor.Request) (tikvpb.Tikv_CoprocessorStreamClient, error) {
```
Comment (zz-jason) [inline store/tikv/mocktikv/cop_handler_dag.go]: this function is so simply, I think we can remove it.

File: store/tikv/mocktikv/cop_handler_dag.go (hunk 1 above)
Comment (tiancaiamao) [reply store/tikv/mocktikv/cop_handler_dag.go]: Theoretically, all kind of query push down are implemented in **coprocessor**,
but @lamxTyler make the analyze API a little different, I'm not sure the specific format, so leave it for him.
One other thing, analyze request doesn't use streaming yet.

**Comment by shenli** [comment]
@tiancaiamao Please resolve the conflicts.

File: store/tikv/mocktikv/cop_handler_dag.go [hunk 3]
```diff
@@ -103,6 +107,38 @@ func (h *rpcHandler) handleCopDAGRequest(req *coprocessor.Request) *coprocessor.
 	return buildResp(chunks, counts, err)
 }
 
+func (h *rpcHandler) handleCopStream(req *coprocessor.Request) (tikvpb.Tikv_CoprocessorStreamClient, error) {
+	if len(req.Ranges) == 0 {
+		return nil, errors.New("request range is null")
+	}
+	if req.GetTp() != kv.ReqTypeDAG {
+		return nil, errors.Errorf("unsupport request type %d", req.GetTp())
```
Comment (zz-jason) [inline store/tikv/mocktikv/cop_handler_dag.go]: s/unsupport/unsupported/

File: store/tikv/mocktikv/cop_handler_dag.go [hunk 4]
```diff
@@ -337,6 +373,98 @@ func flagsToStatementContext(flags uint64) *stmtctx.StatementContext {
 	return sc
 }
 
+type mockClientStream struct{}
```
Comment (zz-jason) [inline store/tikv/mocktikv/cop_handler_dag.go]: 1. s/mockClientStream/baseMockStreamClient/
2. add comment about this struct.

File: store/tikv/mocktikv/cop_handler_dag.go [hunk 5]
```diff
@@ -337,6 +373,98 @@ func flagsToStatementContext(flags uint64) *stmtctx.StatementContext {
...
+
+func (mockClientStream) Header() (metadata.MD, error) { return nil, nil }
+func (mockClientStream) Trailer() metadata.MD         { return nil }
+func (mockClientStream) CloseSend() error             { return nil }
+func (mockClientStream) Context() goctx.Context       { return nil }
+func (mockClientStream) SendMsg(m interface{}) error  { return nil }
+func (mockClientStream) RecvMsg(m interface{}) error  { return nil }
+
+type mockCoprocessorStreamClient struct {
+	mockClientStream
```
Comment (zz-jason) [inline store/tikv/mocktikv/cop_handler_dag.go]: add an empty line between `mockClientStream` and the rest fields.

File: store/tikv/mocktikv/cop_handler_dag.go [hunk 6]
```diff
@@ -337,6 +373,98 @@ func flagsToStatementContext(flags uint64) *stmtctx.StatementContext {
...
+	req      *tipb.DAGRequest
+	exec     executor
+	goCtx    goctx.Context
+	finished bool
+}
+
+type streamRegionError struct {
+	mockClientStream
```
Comment (zz-jason) [inline store/tikv/mocktikv/cop_handler_dag.go]: ditto

File: store/tikv/mocktikv/cop_handler_dag.go [hunk 7]
```diff
@@ -337,6 +373,98 @@ func flagsToStatementContext(flags uint64) *stmtctx.StatementContext {
...
+
+type mockCoprocessorStreamClient struct {
+	mockClientStream
+	req      *tipb.DAGRequest
+	exec     executor
+	goCtx    goctx.Context
+	finished bool
+}
+
+type streamRegionError struct {
```
Comment (zz-jason) [inline store/tikv/mocktikv/cop_handler_dag.go]: s/streamRegionError/mockCoprocessorStreamClient/mockCopStreamErrClient/

File: store/tikv/mocktikv/cop_handler_dag.go [hunk 8]
```diff
@@ -337,6 +373,98 @@ func flagsToStatementContext(flags uint64) *stmtctx.StatementContext {
...
+type mockClientStream struct{}
+
+func (mockClientStream) Header() (metadata.MD, error) { return nil, nil }
+func (mockClientStream) Trailer() metadata.MD         { return nil }
+func (mockClientStream) CloseSend() error             { return nil }
+func (mockClientStream) Context() goctx.Context       { return nil }
+func (mockClientStream) SendMsg(m interface{}) error  { return nil }
+func (mockClientStream) RecvMsg(m interface{}) error  { return nil }
+
+type mockCoprocessorStreamClient struct {
```
Comment (zz-jason) [inline store/tikv/mocktikv/cop_handler_dag.go]: s/mockCoprocessorStreamClient/mockCopStreamClient/

File: store/tikv/mocktikv/cop_handler_dag.go [hunk 9]
```diff
@@ -337,6 +373,98 @@ func flagsToStatementContext(flags uint64) *stmtctx.StatementContext {
...
+	*errorpb.Error
+}
+
+func (e *streamRegionError) Recv() (*coprocessor.Response, error) {
+	return &coprocessor.Response{
+		RegionError: e.Error,
+	}, nil
+}
+
+func (mock *mockCoprocessorStreamClient) Recv() (*coprocessor.Response, error) {
```
Comment (zz-jason) [inline store/tikv/mocktikv/cop_handler_dag.go]: s/mock/e/

File: store/tikv/mocktikv/cop_handler_dag.go [hunk 10]
```diff
@@ -337,6 +373,98 @@ func flagsToStatementContext(flags uint64) *stmtctx.StatementContext {
...
+	data1, err1 := proto.Marshal(&streamResponse)
+	if err1 != nil {
+		resp.OtherError = err1.Error()
+		return &resp, nil
+	}
+	resp.Data = data1
+	return &resp, nil
+}
+
+func readBlockFromExecutor(goCtx goctx.Context, e executor, rowCols []uint32) (tipb.Chunk, bool, error) {
```
Comment (zz-jason) [inline store/tikv/mocktikv/cop_handler_dag.go]: can this be a member function of `mockCoprocessorStreamClient` ?

File: store/tikv/mocktikv/cop_handler_dag.go [hunk 11]
```diff
@@ -337,6 +373,98 @@ func flagsToStatementContext(flags uint64) *stmtctx.StatementContext {
...
+	var err error
+	var count int
+	var chunk tipb.Chunk
+	for count = 0; count < rowsPerChunk; count++ {
```
Comment (zz-jason) [inline store/tikv/mocktikv/cop_handler_dag.go]: s/count = 0/count := 0/, we can remove `var count int`

File: store/tikv/mocktikv/cop_handler_dag.go [hunk 12]
```diff
@@ -337,6 +373,98 @@ func flagsToStatementContext(flags uint64) *stmtctx.StatementContext {
...
+		var row [][]byte
+		row, err = e.Next(goCtx)
```
Comment (zz-jason) [inline store/tikv/mocktikv/cop_handler_dag.go]: s/row, err = e.Next(goCtx)/row, err := e.Next(goCtx)/, we can remove `var err error` and `var row [][]byte`

File: store/tikv/mocktikv/cop_handler_dag.go (hunk 4 above)
Comment (tiancaiamao) [reply store/tikv/mocktikv/cop_handler_dag.go]: Go has no concept of inherit, so we don't need `baseMockStreamClient`.

**Comment by tiancaiamao** [comment]
@zimulala

File: store/tikv/mocktikv/cop_handler_dag.go [hunk 13]
```diff
@@ -104,6 +108,38 @@ func (h *rpcHandler) handleCopDAGRequest(req *coprocessor.Request) *coprocessor.
...
+	ctx := &dagContext{
+		dagReq:    dagReq,
+		keyRanges: req.Ranges,
+		evalCtx:   &evalContext{sc: sc, timeZone: timeZone},
+	}
+	e, err := h.buildDAG(ctx, dagReq.Executors)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
```
Comment (zimulala) [inline store/tikv/mocktikv/cop_handler_dag.go]: Can line112-135 be extracted into a function? It seems the same as the function of `handleCopDAGRequest`.

File: store/tikv/mocktikv/cop_handler_dag.go [hunk 14]
```diff
@@ -370,6 +406,98 @@ func flagsToStatementContext(flags uint64) *stmtctx.StatementContext {
...
+	resp.Data = data1
+	return &resp, nil
+}
+
+func (mock *mockCopStreamClient) readBlockFromExecutor() (tipb.Chunk, bool, error) {
+	var chunk tipb.Chunk
+	for count := 0; count < rowsPerChunk; count++ {
+		row, err := mock.exec.Next(mock.goCtx)
+		if err != nil {
+			return chunk, false, errors.Trace(err)
```
Comment (zimulala) [inline store/tikv/mocktikv/cop_handler_dag.go]: Returning nil is better?

File: store/tikv/mocktikv/cop_handler_dag.go [hunk 15]
```diff
@@ -370,6 +406,98 @@ func flagsToStatementContext(flags uint64) *stmtctx.StatementContext {
...
+	if err != nil {
+		resp.OtherError = err.Error()
+		return &resp, nil
+	}
+	streamResponse := tipb.StreamResponse{
+		Error:      toPBError(err),
+		EncodeType: tipb.EncodeType_TypeDefault,
+		Data:       data,
+	}
+	data1, err1 := proto.Marshal(&streamResponse)
```
Comment (zimulala) [inline store/tikv/mocktikv/cop_handler_dag.go]: We can use "resp.Data" directly.

File: store/tikv/mocktikv/cop_handler_dag.go [hunk 16]
```diff
@@ -370,6 +406,98 @@ func flagsToStatementContext(flags uint64) *stmtctx.StatementContext {
...
+			resp.OtherError = err.Error()
+		}
+		return &resp, nil
+	}
+	if finish {
+		// Just mark it, need to handle the last chunk.
+		mock.finished = true
+	}
+
+	data, err := chunk.Marshal()
```
Comment (zimulala) [inline store/tikv/mocktikv/cop_handler_dag.go]: Can we return chuck directly like `SelectResponse`? Or we can return chuck.RowsData. Then we can reduce the time of `Marshal`.

Comment (tiancaiamao) [reply store/tikv/mocktikv/cop_handler_dag.go]: No, we can't change the data format.
This data format is the result after discussing with TiKV team, mock tikv keep consistent to TiKV.

File: store/tikv/mocktikv/cop_handler_dag.go (hunk 14 above)
Comment (tiancaiamao) [reply store/tikv/mocktikv/cop_handler_dag.go]: The return type is `tipb.Chunk`, not `*tipb.Chunk`, so I can't return nil here.
I guess it doesn't make any difference.

**Pull request merged**